cEP-0025.md

```diff
@@ -0,0 +1,305 @@
+
+Here is a brief overview of the architecture:
+
+1. There will be a separate repository named as coala-pyflakes which will
+   be installable via `pip install pyflakes-bears`.
+2. The repository will contain a new package `pyflakes-bears` which
+   would house all bears which use pyflakes-enhanced-ast.
+3. The `pyflakes-bears` would contain a metabear `PyFlakesASTBear`
+   which would be used by all other plugin bears.
+4. The repository would even contain another package
+   `pyflakes-generic-plugins` that would be independent of coala and could
+   be run by flake8 or coala or any other similar tool.
+
+## Management of the new `coala-pyflakes` repository
+
+The repository will use a similar mechanism of what is being
+used in `coala-bears` to test all its bears.
+
+## Implementation of PyFlakesASTBear
+
+1. `PyFlakesASTBear` will be `LocalBear` returning result of the type
+   `HiddenResult` since these results are not meant for the users but
+   for the dependent bear. The result structure is described by
+   `PyFlakesResult` class.
+2. `PyFlakesResult` consists of a snapshot of the module at class,
+   function, module, generator and doctest level. Thus, providing
+   the developer multiple views of a file allowing him to retrieve
+   results as per his demand.
+3. Apart from scope information, the metabear also returns a list of
+   **messages generated by pyflakes** itself. These s can be used
+   by the developer to add an additional filter layer that is
+   required to be resolved by the user before executing the plugin.
+   This is primarily done because pyflakes emits additional syntax
+   errors which are not caught by python interpretor and may affect
+   the working of a plugin. This check is optional and may or may not
+   be incorparated by the developer.
+4. Helper function `get_node_location` provides with node location
+   in source file.
+5. Helper function `get_nodes` provides with all the nodes of a
+   particular type present in a scope.
+
+Here is a prototype for the implementations of `PyFlakesASTBear`:
+
+```python
+# Imports not mentioned to maintain brevity
+
+class PyFlakesResult(HiddenResult):
+
+    def __init__(self, origin, deadScopes, pyflakes_messages):
+
+        Result.__init__(self, origin, message='')
+
+        self.module_scope = self.get_scopes(ModuleScope, deadScopes)[0]
+        self.class_scopes = self.get_scopes(ClassScope, deadScopes)
+        self.function_scopes = self.get_scopes(FunctionScope, deadScopes)
+        self.generator_scopes = self.get_scopes(GeneratorScope, deadScopes)
+        self.doctest_scopes = self.get_scopes(DoctestScope, deadScopes)
+        self.pyflakes_messages = pyflakes_messages
+
+    def get_scopes(self, scope_type, scopes):
+        return list(filter(lambda scope: isinstance(scope, scope_type),
+                           scopes))
+
+    def get_node_location(self, node):
+        return {'lineno': node.source.lineno,
+                'depth': node.source.depth - 1,
```

Comment:
`depth` isnt used in your examples.

and this method is just a thin wrapper around `node.source` ... maybe it should return `node.source`, which allows access to `lineno` and `col_offset`, and if pyflakes internals ever change then get_node_location` could return its own `NamedTuple` which provides those attributes.